### PR TITLE
 feat (audience match types): Update condition evaluator for new audience match types.

### PR DESCRIPTION
**Pull request by rashidsp**
**Summary**
This adds support for new audience match type conditions to the condition evaluator:

exact, exists, gt, lt, and substring conditions
Abort and return null when appropriate in leaf evaluators
Null handling in and/or/not evaluators
Assume implicit "or" when operator is missing

### CODE REVIEW

**Review by oakbani — CHANGES_REQUESTED** [review]
Reviewed Unit Tests. Will review code after Prayer

File: lib/optimizely/audience.rb [hunk 1]
```diff
@@ -16,7 +16,8 @@
 #    limitations under the License.
 #
 require 'json'
-require_relative './condition'
+require_relative './condition_tree_evaluator'
```
[the remark this answers is not in the document]
Comment (rashidsp) [reply lib/optimizely/audience.rb]: Addressed in https://github.com/msohailhussain/ruby-sdk/pull/50/files#diff-e3f83db360101dcd3cf4247456af6b14L4

File: lib/optimizely/custom_attribute_condition_evaluator.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                    different_types?(condition_type, user_provided_type)
+
+      condition_value == user_provided_value
+    end
+
+    def exists_evaluator(condition)
+      # Evaluate the given exists match condition for the given user attributes.
+      # Returns boolean true if both:
+      #                    1) the user attributes have a value for the given condition, and
+      #                    2) the user attribute value is neither null nor undefined
```
Comment (oakbani) [inline lib/optimizely/custom_attribute_condition_evaluator.rb]: Remove undefined. undefined is a type in JS, not in ruby

File: lib/optimizely/custom_attribute_condition_evaluator.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    private
+
+    def value_valid_for_exact_conditions?(value)
+      # Returns true if the value is valid for exact conditions. Valid values include
+      #  strings, booleans, and numbers that aren't NaN, -Infinity, or Infinity.
+
+      return value.finite? if value.is_a? Float
+
+      (value.is_a? FalseClass) || (value.is_a? Integer) || (value.is_a? String) ||
```
Comment (oakbani) [inline lib/optimizely/custom_attribute_condition_evaluator.rb]: don't we need a finite check with integer class when we need it for float?

File: lib/optimizely/custom_attribute_condition_evaluator.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      user_provided_value.include? condition_value
+    end
+
+    private
+
+    def value_valid_for_exact_conditions?(value)
+      # Returns true if the value is valid for exact conditions. Valid values include
+      #  strings, booleans, and numbers that aren't NaN, -Infinity, or Infinity.
+
+      return value.finite? if value.is_a? Float
```
Comment (oakbani) [inline lib/optimizely/custom_attribute_condition_evaluator.rb]: Reuse def infinite_number?(value)

File: lib/optimizely/custom_attribute_condition_evaluator.rb [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+      (value.is_a? FalseClass) || (value.is_a? Integer) || (value.is_a? String) ||
+        (value.is_a? TrueClass)
+    end
+
+    def different_types?(condition_type, user_provided_type)
```
Comment (oakbani) [inline lib/optimizely/custom_attribute_condition_evaluator.rb]: I don't get the need of this method. Exact should check that both user and condition values need to have same type. Why specific boolean check?

**Review by oakbani — CHANGES_REQUESTED** [review]
Code review done

File: lib/optimizely/custom_attribute_condition_evaluator.rb (hunk 3 above)
Comment (rashidsp) [reply lib/optimizely/custom_attribute_condition_evaluator.rb]: Ruby throws ZeroDivisionError against integer division.
Class infinity works with float.

File: lib/optimizely/custom_attribute_condition_evaluator.rb (hunk 4 above)
Comment (rashidsp) [reply lib/optimizely/custom_attribute_condition_evaluator.rb]: Float check will be redundant.

File: lib/optimizely/custom_attribute_condition_evaluator.rb (hunk 5 above)
Comment (rashidsp) [reply lib/optimizely/custom_attribute_condition_evaluator.rb]: In ruby true and false have different type classes, TrueClass and FalseClass respectively.
Therefore there is need to check explicitly.

File: lib/optimizely/custom_attribute_condition_evaluator.rb (hunk 2 above)
Comment (rashidsp) [reply lib/optimizely/custom_attribute_condition_evaluator.rb]: undefined is correct and updated null to nil.

**Comment by rashidsp** [comment]
Creating new PR

**Pull request closed without merging**